### PR TITLE
Add Ecmp Static Route for centralized gateway

### PR DESCRIPTION
**Pull request by hongzhen-ma**

### CODE REVIEW

File: test/e2e/subnet/normal.go [hunk 1]
```diff
@@ -117,7 +117,7 @@ var _ = Describe("[Subnet]", func() {
 			subnet, err := f.OvnClientSet.KubeovnV1().Subnets().Get(context.Background(), name, metav1.GetOptions{})
 			Expect(err).NotTo(HaveOccurred())
 			Expect(subnet.Spec.GatewayType).To(Equal(kubeovn.GWCentralizedType))
-			Expect(subnet.Status.ActivateGateway).To(Equal("kube-ovn-control-plane"))
+			// Expect(subnet.Status.ActivateGateway).To(Equal("kube-ovn-control-plane"))
```
Comment (oilbeater) [inline test/e2e/subnet/normal.go:120]: No need to remain the codes here

File: pkg/ovs/ovn-nbctl.go [hunk 2]
```diff
@@ -463,7 +463,7 @@ func (c Client) ListStaticRoute() ([]StaticRoute, error) {
 }
 
 // AddStaticRoute add a static route rule in ovn
-func (c Client) AddStaticRoute(policy, cidr, nextHop, router string) error {
+func (c Client) AddStaticRoute(policy, cidr, nextHop, router string, isCentralized bool) error {
```
Comment (oilbeater) [inline pkg/ovs/ovn-nbctl.go:466]: `isCentralized` -> `ecmp` for better understanding

File: pkg/controller/subnet.go [hunk 3]
```diff
@@ -849,15 +849,15 @@ func (c *Controller) reconcileGateway(subnet *kubeovnv1.Subnet) error {
 	} else {
 		// if gw is distributed remove activateGateway field
 		if subnet.Spec.GatewayType == kubeovnv1.GWDistributedType {
-			if subnet.Status.ActivateGateway == "" {
+			if subnet.Spec.GatewayNode == "" {
```
Comment (oilbeater) [inline pkg/controller/subnet.go:852]: why not update the ActivateGateway

File: pkg/controller/subnet.go [hunk 4]
```diff
@@ -887,56 +887,46 @@ func (c *Controller) reconcileGateway(subnet *kubeovnv1.Subnet) error {
...
-				subnet.Status.NotReady("NoReadyGateway", "")
-				bytes, err := subnet.Status.Bytes()
-				if err != nil {
-					return err
-				}
-				_, err = c.config.KubeOvnClient.KubeovnV1().Subnets().Patch(context.Background(), subnet.Name, types.MergePatchType, bytes, metav1.PatchOptions{}, "status")
-				return err
+			nodeIPs, err = c.filterEcmpStaticRoute(nodeIPs, subnet.Spec.CIDRBlock)
+			if err != nil {
+				klog.Errorf("add ecmp static route for subnet %v, error %v", subnet.Name, err)
```
Comment (oilbeater) [inline pkg/controller/subnet.go:924]: add -> filter

File: pkg/controller/subnet.go [hunk 5]
```diff
@@ -1192,3 +1181,31 @@ func filterRepeatIPRange(mapIps map[string]ipam.IPRange) map[string]ipam.IPRange
...
+	if len(nodeIps) == 0 {
+		return nextHops, fmt.Errorf("nexthop is nil for add ecmp static route")
+	}
+
+	for _, route := range routes {
+		if route.Policy != ovs.PolicySrcIP {
+			continue
+		}
+		for _, nodeIp := range nodeIps {
+			if route.CIDR == cidrBlock && route.NextHop == nodeIp {
```
Comment (oilbeater) [inline pkg/controller/subnet.go:1201]: what happens if insert duplicated ecmp routes

File: pkg/controller/subnet.go [hunk 6]
```diff
@@ -1192,3 +1181,31 @@ func filterRepeatIPRange(mapIps map[string]ipam.IPRange) map[string]ipam.IPRange
 	}
 	return mapIps
 }
+
+func (c *Controller) filterEcmpStaticRoute(nodeIps []string, cidrBlock string) ([]string, error) {
```
Comment (oilbeater) [inline pkg/controller/subnet.go:1185]: use a better understanding name

File: pkg/controller/node.go [hunk 7]
```diff
@@ -225,7 +225,7 @@ func (c *Controller) handleAddNode(key string) error {
 	nodeAddr := util.GetNodeInternalIP(*node)
 	for _, ip := range strings.Split(ipStr, ",") {
 		if util.CheckProtocol(nodeAddr) == util.CheckProtocol(ip) {
-			err = c.ovnClient.AddStaticRoute("", nodeAddr, ip, c.config.ClusterRouter)
+			err = c.ovnClient.AddStaticRoute("", nodeAddr, ip, c.config.ClusterRouter, false)
```
Comment (oilbeater) [inline pkg/controller/node.go:228]: use const or type to define use ecmp or not for better understanding

**Review by oilbeater — CHANGES_REQUESTED** [review]

**Comment by oilbeater** [comment]
For multiple centralized gateway node, we need to detect the node healthy status and update the routes for high available @hongzhen-ma

File: test/e2e/subnet/normal.go (hunk 1 above)
Comment (hongzhen-ma) [reply test/e2e/subnet/normal.go]: modified

File: pkg/controller/subnet.go (hunk 6 above)
Comment (hongzhen-ma) [reply pkg/controller/subnet.go]: modified

File: pkg/controller/node.go (hunk 7 above)
Comment (hongzhen-ma) [reply pkg/controller/node.go]: modified

**Review by oilbeater — APPROVED** [review]

**Pull request merged**